### ISSUE TITLE
Handle assertions as exceptions and return an error message instead of termination in `osrm-routed`

### ISSUE THREAD

**Issue by oxidase** [Project-OSRM/osrm-backend#3559]
If `osrm-routed` is compiled with assertions than the current implementation will terminate at https://github.com/Project-OSRM/osrm-backend/blob/master/src/util/assert.cpp#L14

It would be good to switch to exceptions, catch an assertion at https://github.com/Project-OSRM/osrm-backend/blob/master/src/server/request_handler.cpp#L65 and possibly return a correct HTTP error answer with a JSON debug message.

/cc @daniel-j-h @MoKob

**Comment by daniel-j-h** [comment]
Yes please - this is a huge pain point at the moment where the demo server crashes every 5 minutes.
